### PR TITLE
[otbn,util] Halve the DMEM region in OTBN's linker script

### PR DESCRIPTION
**Pull request by rswarbrick**
This is to match how only the bottom half of DMEM is bus-accessible.

### CODE REVIEW

**Comment by rswarbrick** [comment]
This ain't gonna work without a bit of fiddling, because `rsa.s` genuinely has more than 2kb of data that needs to be bus-accessible. Oops! Fortunately, two of the buffers can be merged together: see #9247.

**Comment by GregAC** [comment]
> This ain't gonna work without a bit of fiddling, because `rsa.s` genuinely has more than 2kb of data that needs to be bus-accessible. Oops! Fortunately, two of the buffers can be merged together: see #9247.

We do need to be careful the 2kb won't cause us problems later on. So it's a little concerning we've run into an issue with RSA, albeit a trivially resolved one.

I guess if we do hit something that ends up needing more than 2kb of bus accessible data we can work around it with little programs that copy from bus accessible area to otbn only area and vice versa, a bit awkward though! @felixmiller any thoughts?

**Comment by rswarbrick** [comment]
> > This ain't gonna work without a bit of fiddling, because `rsa.s` genuinely has more than 2kb of data that needs to be bus-accessible. Oops! Fortunately, two of the buffers can be merged together: see #9247.
> 
> We do need to be careful the 2kb won't cause us problems later on. So it's a little concerning we've run into an issue with RSA, albeit a trivially resolved one.
> 
> I guess if we do hit something that ends up needing more than 2kb of bus accessible data we can work around it with little programs that copy from bus accessible area to otbn only area and vice versa, a bit awkward though! @felixmiller any thoughts?

I know what you mean. I probably suggest merging something like this (assuming it works!) for now. If we need to change the fraction of DMEM that's visible, it will be a small matter of tweaking some constants, but the infrastructure won't need to change.

**Review by GregAC — APPROVED** [review]

**Comment by felixmiller** [comment]
Some thoughts on this:

**TL;DR:** Personally, I would feel most comfortable if we could delay the decision for an exact window size a bit. And then decide between either 1ki or 2ki size for the protected area.

The RSA situation is probably the one which requires the largest share of bus accessible DMEM from all crypto operations. Actually, was thinking of that last time we discussed, but didn't think of the in/out situation @rswarbrick ran into, so I concluded 2k will be good. The same issue could also come up in the context of RSA verify where we do not need the exponent but might need to supply the RR constant (which is key size) if it is not online computed. So also here it would be 3x512 Byte + X for RSA 4k (where X are some smaller parameters definitely smaller than 512 Byte). So I would say for RSA we are good with 2k of accessible DMEM. Having a "working buffer" outside of the bus visible area is actually a good thing, since intermediate results are not available on the bus anymore then.

The other critical application would be a generic ECC lib where we would have to supply all the domain parameters via the bus (in addition to the actual parameters of the crypto operation).

Currently we the widest curve we support is P-384. Thinking of a worst case scenario this would be even wider curves e.g. P-521 and everything being 256-bit aligned such that a scalar and a single coordinate consumes 768 bit each. Summary of domain parameters for curves in Short Weierstrass shape would be then:
- a: scalar size =>768 bit => 96 Byte
- b: scalar size => 768 bit => 96 Byte
- G: curve point => worst case in uncompressed i.e. 3 coordinates each => 288 Byte
- h: scalar size=> 96 Byte
- p: scalar size => 96 Byte
=> 768 Byte in total for domain parameters (for P-384 this would be half of this in the worst case).

In addition, for ECDSA we need the parameters for the crypto operations which are

- for signature creation: R,S,d,m,k + blinding => 576 Byte
- for signature verification: S, Q (curve point), k, res + blinding => 672 Byte
The above would be absolute worst case. Not considering any buffers that could be shared, assuming the points to be provided in 3-coordinate form and assuming that we supply the nonce and blinding params via the bus.
So the absolute worst case for anything ECC related I can come up with would be 1440 Bytes.

Another problematic case could be if we ever decide to do the hashing on OTBN. But I guess in this case, it would be a multi  step thing with an "update" function anyways.

**On the other hand:** The reason why we came up with this idea initially, are crypto operations where ephemeral keys are generated (mainly ECIES). In any case, the algorithms we would use there would be ECC based, meaning fairly small keys to be hidden in the non-accessibly window. Even for P-521 and if there are multiple keys, 1k of non-accessible space would be enough.

Personally, I would feel most comfortable if we could delay the decision for an exact window size a bit. I guess having a 1ki "protected" window would not make any difference in terms of logic complexity. I think we can rule out anything other than 1ki or 2ki.

**Comment by rswarbrick** [comment]
You're right, there's no real difficulty in adjusting the sizes. At the moment, we've updated the documentation for 2kb in each half. If it's ok with the two of you, I propose to merge this PR and the design PR that implements things. Fixing everything up to switch to 3kb + 1kb is just a matter of fiddling around with some magic numbers, but it would probably be easier to do after the fact (and then we'll get a nice "now resize everything" commit).

**Comment by rswarbrick** [comment]
Wow! Git is awesome! This rebased over the big file rename from #9305 with no manual intervention!
